docs/development/test-guidelines.adoc

```diff
@@ -0,0 +1,209 @@
+= Test Guidelines
+:toc:
+:toc-placement: preamble
+
+:sectnums:
+:experimental:
+
+This documents contains various information on testing the codebase. Over time it might
+grow to include sections on other testing strategies than unit or integration tests. 
+
+[[how-to-write-unit-tests]]
+== How to write unit tests?
+
+Instead of writing down general rules for how to write tests, this section presents a
+step-by-step approach on how to write tests for a very simply library file called `lib.go`
+that lives inside the `foo/bar` package.
+The library contains three one-line functions which means that we have to test three functions
+in order to gain 100% test coverage. For simplicity, the functions under test don't have
+any dependency on an external resource like a database connection or something else.
+Nevertheless we will present mechanisms to handle this as well.
+
+[[lib-dot-go-source]]
+[source,go]
+.$GOPATH/src/foo/bar/lib.go
+----
+package bar // <1>
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Say returns the input string as is
+func Say(message string) string { // <2>
+	return fmt.Sprintf("%s", message)
+}
+
+// Shout returns the input string in upper case
+// letters with three exlamation marks.
+func Shout(message string) string { // <3>
+	return fmt.Sprintf("%s!!!", toUpper(message))
+}
+
+// toUpper returns the input string in upper case
+func toUpper(message string) string { // <4>
+	return strings.ToUpper(message)
+}
+----
+<1> The package name is `bar`. 
+<2> An exported function called `Say` that is unknown packages that import `bar`.
+<3> Another exported function called `Shout` that is unknown packages that import `bar`.
+<4> A function called `toUpper` that is unexported and unknown packages that import `bar`.
+
+=== Blackbox vs. Whitebox tests
+
+In a blackbox test, you test a package as if you **only have access to the
+exported symbols**. In our <<lib-dot-go-source>> example above we have two exported
+symbols: `Say` and `Shout`. The `toUpper` function is an unexported symbol due
+to its lowercase name. 
+
+A whitebox test has access to all symbols, and not only to the expored ones.
+
+In principle, all symbols in a file can be *seen* by all files in the same package.
+As a consequence, to have a blackbox test, we must write it in a different package
+than `bar`. But in order for `go test` to pick up a test we need to place it in the
+same directory as the <<lib-dot-go-source>> file.
+
+In Go your test filenames must end with `_test.go`. In order to distinguish black-
+and whitebox test files, a recommendation is to use the suffixes `_blackbox_test.go`
+and `_whitebox_test.go`.
```

Comment:
This makes for long filenames. As @baijum (I think) suggested, we could have whitebox as the default and call blackbox tests "<sth>_bb_test.go". If we establish this as a pattern, people will know what it means.